primitives/runtime/src/offchain/pollable.rs

```diff
@@ -0,0 +1,439 @@
+		let mut block = true;
+
+		// We process in a loop all pending messages.
+		while let Some((message_id, index_in_list)) = next_notification(&mut state.message_ids, block) {
+			block = false;
+
+			state.message_ids.remove(index_in_list);
+			let waker = state.wakers.remove(index_in_list);
+			waker.wake();
+
+			let _new_msg_id = state.pending_messages.insert(message_id);
+			debug_assert!(_new_msg_id);
+		}
+
+		debug_assert!(!block);
+	}
+}
+
+lazy_static::lazy_static! {
+	// TODO: we're using a Mutex, which is ok for as long as WASM doesn't have threads
+	// if WASM ever gets threads and no pre-emptive multitasking, then we might spin forever
+	static ref STATE: Mutex<BlockOnState> = {
+		Mutex::new(BlockOnState {
+			message_ids: Vec::new(),
+			wakers: Vec::new(),
+			pending_messages: BTreeSet::new(),
+		})
+	};
+}
+
+/// State of the global `block_on` mechanism.
+///
+/// This is instantiated only once.
+struct BlockOnState {
+	/// List of messages for which we are waiting for a response. A pointer to this list is passed
+	/// to the kernel.
+	message_ids: Vec<MessageId>,
+
+	/// List whose length is identical to [`BlockOnState::messages_ids`]. For each element in
+	/// [`BlockOnState::messages_ids`], contains a corresponding `Waker` that must be waken up
+	/// when a response comes.
+	wakers: Vec<Waker>,
+
+	/// Queue of response messages waiting to be delivered.
+	///
+	/// > **Note**: We have to maintain this queue as a global variable rather than a per-future
+	/// >           channel, otherwise dropping a `Future` would silently drop messages that have
+	/// >           already been received.
+	pending_messages: BTreeSet<MessageId>,
+}
+
+/// Future that's resolved whenever the `PollableId` is signalled as ready by
+/// the Substrate host.
+#[must_use = "futures do nothing unless polled"]
+#[derive(Debug)]
+pub struct HostFuture {
+	msg_id: MessageId,
+	finished: bool,
+}
+
+impl<T> From<T> for HostFuture where T: Into<PollableId> {
+	fn from(value: T) -> HostFuture {
+		Self {
+			msg_id: value.into(),
+			finished: false,
+		}
+	}
+}
+
+impl Future for HostFuture {
+	type Output = ();
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+		assert!(!self.finished);
+		if peek_response(self.msg_id) {
+			self.finished = true;
+			Poll::Ready(())
+		} else {
+			register_message_waker(self.msg_id, cx.waker().clone());
+			Poll::Pending
+		}
+	}
+}
+
+/// If a response to this message ID has previously been obtained, extracts it for processing.
+pub(crate) fn peek_response(msg_id: MessageId) -> bool {
+	let mut state = STATE.lock();
+	state.pending_messages.remove(&msg_id)
+}
+
+/// Future that resolves to a HTTP request response.
+///
+/// Because HTTP response can be chunked, the resolved [`HttpResponse`] contains
+/// both HTTP status code and headers but the [`HttpResponse::body`] is a future
+/// type that has to be driven separately.
+#[pin_project]
```

Comment:
Is it necessary to use `pin_project`? None of the futures will have state anyway, so they are probably `Unpin`. Just trying to avoid adding unnecessary libs to already big wasm runtime.

Reply:
You're right, it was more of a reflex :) Fixed!